File: behavioral/memento.php
<?php
class Editor {
    private string $text = "";
    public function type(string $words){ $this->text .= " " . $words; }
    public function getText(): string { return $this->text; }
    public function save(): Memento { return new Memento($this->text); }
    public function restore(Memento $m){ $this->text = $m->getText(); }
}

class Memento {
    private string $text;
    public function __construct(string $t){ $this->text = $t; }
    public function getText(): string { return $this->text; }
}

// تست
$e = new Editor();
$e->type("salam");
$saved = $e->save();
$e->type("chetori?");
echo $e->getText()."\n";   // salam chetori?
$e->restore($saved);
echo $e->getText()."\n";   // salam

File: behavioral/memento.go
package main

import "fmt"

type Memento struct{ text string }

func (m Memento) GetText() string { return m.text }

type Editor struct{ text string }

func (e *Editor) Type(t string)     { e.text += " " + t }
func (e *Editor) GetText() string   { return e.text }
func (e *Editor) Save() Memento     { return Memento{text: e.text} }
func (e *Editor) Restore(m Memento) { e.text = m.GetText() }

func main() {
	e := &Editor{}
	e.Type("Salam")
	saved := e.Save()
	e.Type("chetori?")
	fmt.Println(e.GetText())
	e.Restore(saved)
	fmt.Println(e.GetText())
}
